### DISCUSSION TITLE
How to Create Composable Settings with Pydantic?

### DISCUSSION THREAD

**Discussion by juskaiser** [pydantic/pydantic#8353] [Question]
I would like to use Pydantic to create composable settings, where elements from environment-based settings and country-based settings can be combined into a final settings object.

This is a non-working example that demonstrates what I'd like to accomplish:
```python
from pydantic import BaseSettings


class S3Settings(BaseSettings):
    """Settings for S3."""

    bucket: str  # This config is based on environment
    market: str  # This config is based on the country/market
    project_path: str = "my_project"

    @property
    def base_path(self: "S3Settings") -> str:
        """Get the base path for the project in S3."""
        return f"s3://{self.bucket}/{self.project_path}/{self.market}"


class CommonSettings(BaseSettings):
    """Shared configuration for all environments."""
    market: str
    s3: S3Settings


class StagingCommonSettings(CommonSettings):
    """Configuration for staging environment."""
    market: str
    s3 = S3Settings(bucket="staging", market=market)


class ProdCommonSettings(CommonSettings):
    """Configuration for production environment."""
    market: str
    s3 = S3Settings(bucket="prod", market=market)


class USCommonSettings(BaseSettings):
    market = "US"


# Inherits from both of the parent classes. The value for `market` should come from USCommonSettings, 
#    and the value for `bucket` should come from StagingCommonSettings and ProdCommonSettings respectively.
class USStagingSettings(USCommonSettings, StagingCommonSettings):
    pass


class USProdSettings(USCommonSettings, ProdCommonSettings):
    pass
```
This type of structure would allow defaults that get inherited for each environment and country. It would also allow for specific overrides as needed in the composite `USStagingSettings` and `USProdSettings` classes.

Currently, when I try to import and instantiate `USStagingSettings` I get an error from the `StagingCommonSettings` that `market` is not defined. I was hoping that by with the Muli-Parent class structure, `market` would get set by `USCommonSettings` and would then be referenced by `StagingCommonSettings` as the class is constructed.

Does anyone have suggestions or examples on how to implement a pattern like this? Thanks in advance for any help!

Additional Notes:
> I'm currently using Pydantic ` 1.10.13` and probably cannot update to 2.x due to constraints of some of the other packages I'm using.

**Comment by adrianog** [comment]
did you sort this?
